test/test_overrides.py

```diff
@@ -0,0 +1,756 @@
+        if type(other) is type(self):
+            if self._N == other._N and self._i == other._i:
+                return True
+            else:
+                return False
+        else:
+            return False
+
+@six.add_metaclass(ImplementationMeta)
+class SubTensor(torch.Tensor):
+    """A subclass of torch.Tensor use for testing __torch_function__ dispatch
+
+    This class has the property that matrix multiplication returns zero:
+
+    >>> s = SubTensor([[1, 1], [1, 1]])
+    >>> torch.mm(s, s)
+    0
+    >>> t = torch.tensor([[1, 1], [1, 1]])
+    >>> torch.mm(s, t)
+    0
+    >>> torch.mm(t, s)
+    0
+    >>> torch.mm(t, t)
+    tensor([[2, 2],
+            [2, 2]])
+
+    This is useful for testing that the semantics for overriding torch
+    functions are working correctly
+
+    """
+    TORCH_IMPLEMENTATIONS = (
+        (torch.mm, lambda mat1, mat2: 0),
+    )
+    IMPLEMENTS_DECORATOR = implements_sub
+
+    def __torch_function__(self, func, args=(), kwargs=None):
+        if(kwargs is None):
+            kwargs = {}
+
+        if func not in HANDLED_FUNCTIONS_SUB:
+            return NotImplemented
+        # Note: this allows subclasses that don't override
+        # __torch_function__ to handle DiagonalTensor objects.
+        return HANDLED_FUNCTIONS_SUB[func](*args, **kwargs)
+
+@six.add_metaclass(ImplementationMeta)
+class SubDiagonalTensor(DiagonalTensor):
+    """A subclass of ``DiagonalTensor`` to test custom dispatch
+
+    This class tests semantics for defining ``__torch_function__`` on a
+    subclass of another class that defines ``__torch_function__``. This
+    class provides custom implementations of ``mean`` and ``mm``,
+    scaling the mean by a factor of 10 and returning 1 from ``mm``
+    instead of 0 as ``DiagonalTensor`` does.
+    """
+    TORCH_IMPLEMENTATIONS = (
+        (torch.mean, lambda mat: 10 * float(mat._i) / mat._N),
+        (torch.mm, lambda mat1, mat2: 1),
+    )
+    IMPLEMENTS_DECORATOR = implements_sub_diagonal
+
+    def __init__(self, N, value):
+        self._N = N
+        self._i = value
+
+    def __repr__(self):
+        return "SubDiagonalTensor(N={}, value={})".format(self._N, self._i)
+
+    def __array__(self):
+        return self._i * np.eye(self._N)
+
+    def tensor(self):
+        return self._i * torch.eye(self._N)
+
+    def __torch_function__(self, func, args=(), kwargs=None):
+        if(kwargs is None):
+            kwargs = {}
+
+        if func not in HANDLED_FUNCTIONS_SUB_DIAGONAL:
+            return NotImplemented
+        # In this case _torch_function_ should override DiagonalTensor objects
+        return HANDLED_FUNCTIONS_SUB_DIAGONAL[func](*args, **kwargs)
+
+    def __eq__(self, other):
+        if type(other) is type(self):
+            if self._N == other._N and self._i == other._i:
+                return True
+            else:
+                return False
+        else:
+            return False
+
+class TestOverride(TestCase):
```

Comment:
No need for different test cases?

Reply:
Sorry, to clarify, are you saying that this test and the ones below can be condensed into a single `TestCase` subclass?

Reply:
Yes.